### PR TITLE
Some python improvments

### PR DESCRIPTION
**Pull request by oliver-wm**
Add memoization to umm.py.
Consolidate logic for risc-v and arm memory base.

Signed-off-by: Oliver Scott <Oliver.Scott@data61.csiro.au>

### CODE REVIEW

**Comment by oliver-wm** [comment]
This may or may not be needed, just had them in a working directory and thought it could be useful.

File: tools/hardware/utils/memory.py [hunk 1]
```diff
@@ -61,11 +61,6 @@ def align_memory(regions: Set[Region], config: Config) -> List[Region]:
     ret = sorted(regions)
     extra_reserved = set()
 
-    if config.arch == 'riscv':
-        # RISC-V is special: it expects physBase to be
-        # the address that the bootloader is loaded at.
-        physBase = ret[0].base
```
Comment (kent-mcleod) [inline tools/hardware/utils/memory.py:67]: How come this doesn't need to be set before the following two if statements anymore? It looks like it still needs separate treatment compared to Arm.

Comment (oliver-wm) [reply tools/hardware/utils/memory.py:67]: Not sure this was from a month or so ago. It does look like those ifs are needed, will remove.

Comment (oliver-wm) [reply tools/hardware/utils/memory.py:67]: I've refactored align_memory into the config class. I think this makes it much clearer as to what is going on.

Comment (axel-h) [reply tools/hardware/utils/memory.py:67]: According to @kent-mcleod, this ist kind of RISC-V hack that https://github.com/seL4/seL4/pull/365 will resolve then.

**Review by lsf37 — APPROVED** [review]
This now looks fine to me (style format still needs to be run). Any concerns merging this @kent-mcleod ?

**Comment by lsf37** [comment]
(style fix and rebased)

**Comment by axel-h** [comment]
I'm fin with these changes if they have been tested and work. I have https://github.com/seL4/seL4/pull/365 with more cleanup that I will rebase on top of this then and test that this works with out RISC-V changes

File: tools/hardware/irq.py [hunk 2]
```diff
@@ -185,6 +185,7 @@ def parse_irq(self, child, data):
     'qcom,msm-qgic2': ArmGic,
     'ti,am33xx-intc': RawIrqController,
     'ti,omap3-intc': RawIrqController,
+    'riscv,plic0': RawIrqController,
```
Comment (axel-h) [inline tools/hardware/irq.py:188]: We could make this a separate commit, because it is really separate from the other changes

File: tools/umm.py [hunk 3]
```diff
@@ -76,9 +76,19 @@ def base_name(x):
     return x[1]
 
 # This assumes that membership is a DAG which is the case in C
-# We could memoize, doesn't seem worth it ...
 
 
+def memoize(func):
```
Comment (axel-h) [inline tools/umm.py:81]: So is this worth it now or not?  And when we add this, the comment "# This assumes that membership is a DAG which is the case in C" should be moved down where `paths_to_type` is declared.

Comment (lsf37) [reply tools/umm.py:81]: My initial thought was it doesn't hurt, but you are right, memoization is complete overkill here, the runtime of the entire program is already insignificant, and it just adds unneeded complexity. So, I'll remove it and the comment.

File: tools/hardware/config.py [hunk 4]
```diff
@@ -29,6 +31,12 @@ def get_device_page_bits(self) -> int:
         ''' Get page size in bits for mapping devices for this arch '''
         return self.get_page_bits()
 
+    def align_memory(self, regions: Set[Region]) -> List[Region]:
+        ''' Given a set of regions, sort them and align the first so that the ELF loader will be able to load the kernel into it. Will return the aligned
```
Comment (axel-h) [inline tools/hardware/config.py:35]: Break the line to improve readbility:
```suggestion
        ''' Given a set of regions, sort them and align the first so that the 
        ELF loader will be able to load the kernel into it. Will return the
        aligned memory region list, a set of any regions of memory that were
        aligned out and the physBase value that the kernel will use. '''
```

File: tools/hardware/config.py [hunk 5]
```diff
@@ -39,6 +47,20 @@ def get_kernel_phys_align(self) -> int:
         ''' on ARM the ELF loader expects to be able to map a supersection page to load the kernel. '''
         return self.SUPERSECTION_BITS
 
+    def align_memory(self, regions: Set[Region]) -> List[Region]:
+        ''' Arm wants physBase to be the physical load address of the kernel. '''
```
Comment (axel-h) [inline tools/hardware/config.py:51]: "Arm" -> "ARM"?
```suggestion
        ''' ARM wants physBase to be the physical load address of the kernel. '''
```

File: tools/hardware/config.py [hunk 6]
```diff
@@ -50,6 +72,20 @@ def get_bootloader_reserve(self) -> int:
         of physical memory. Mark it as unavailable. '''
         return self.MEGA_PAGE_SIZE
 
+    def align_memory(self, regions: Set[Region]) -> List[Region]:
+        ''' RISC-V expects physBase to be the address that the bootloader is loaded at. '''
```
Comment (axel-h) [inline tools/hardware/config.py:76]: I have some doubt that this is really what RISC-V wants. This just seem the be something that is currently there due to the history of the RISC-V port - and it feels wrong that we make assumption about the bootloader. If you plan to merge this PR now, it is at least helpful bring some light on the current implementation. But the comment could say this is not the intended long term solution because this creates a nasty dependency on the botloader and how big it is. The other PR (https://github.com/seL4/seL4/pull/365) I'm working at will remove all this code again and use DTB overlays to reserve bootloader memory and align things with how it's done on ARM.

File: tools/hardware/config.py (hunk 5 above)
Comment (lsf37) [reply tools/hardware/config.py:51]: Arm is the correct spelling these days.

File: tools/hardware/config.py (hunk 6 above)
Comment (lsf37) [reply tools/hardware/config.py:76]: Happy to add that

File: tools/hardware/config.py (hunk 5 above)
Comment (lsf37) [reply tools/hardware/config.py:54]: It's annoying that they changed it. I'm not really keen on replacing it everywhere in a PR like this, using the old one seems wrong, and using the new one for new stuff looks inconsistent. Using the latter now and at some point doing a PR that systematically replaces ARM with Arm is probably the way to go.

Comment (kent-mcleod) [reply tools/hardware/config.py:54]: We normally choose to update it if the code is being touched.  This is the same guidelines we follow for style changes too.

Comment (lsf37) [reply tools/hardware/config.py:54]: That makes sense
